server/routes/utils/dataReportHelpers.ts

```diff
@@ -0,0 +1,212 @@
+								esb.matchPhraseQuery(item.meta.key, item.meta.params.query)
+							);
+							break;
+						case 'exists':
+							requestBody.must(esb.existsQuery(item.meta.key));
+							break;
+						case 'phrases':
+							if (item.meta.value.indexOf(',') > -1) {
+								const valueSplit = item.meta.value.split(', ');
+								for (const [key, incr] of valueSplit.entries()) {
+									requestBody.should(esb.matchPhraseQuery(item.meta.key, incr));
+								}
+							} else {
+								requestBody.should(
+									esb.matchPhraseQuery(item.meta.key, item.meta.params.query)
+								);
+							}
+							requestBody.minimumShouldMatch(1);
+							break;
+					}
+					break;
+				case true:
+					switch (item.meta.type) {
+						case 'phrase':
+							requestBody.mustNot(
+								esb.matchPhraseQuery(item.meta.key, item.meta.params.query)
+							);
+							break;
+						case 'exists':
+							requestBody.mustNot(esb.existsQuery(item.meta.key));
+							break;
+						case 'phrases':
+							if (item.meta.value.indexOf(',') > -1) {
+								const valueSplit = item.meta.value.split(', ');
+								for (const [key, incr] of valueSplit.entries()) {
+									requestBody.should(esb.matchPhraseQuery(item.meta.key, incr));
+								}
+							} else {
+								requestBody.should(
+									esb.matchPhraseQuery(item.meta.key, item.meta.params.query)
+								);
+							}
+							requestBody.minimumShouldMatch(1);
+							break;
+					}
+					break;
+			}
+		}
+	}
+	//search part
+	let searchQuery = JSON.parse(filters)
+		.query.query.replace(/ and /g, ' AND ')
+		.replace(/ or /g, ' OR ')
+		.replace(/ not /g, ' NOT ');
+	if (searchQuery) {
+		requestBody.must(esb.queryStringQuery(searchQuery));
+	}
+
+	if (report._source.timeFieldName && report._source.timeFieldName.length > 0) {
+		requestBody.must(
+			esb
+				.rangeQuery(report._source.timeFieldName)
+				.format('epoch_millis')
+				.gte(report._source.start)
+				.lte(report._source.end)
+		);
+	}
+	if (is_count) {
+		return esb.requestBodySearch().query(requestBody);
+	}
+
+	//Add the Sort to the query
+	let reqBody = esb.requestBodySearch().query(requestBody).version(true);
+
+	if (report._source.sorting.length > 0) {
+		if (report._source.sorting.length === 1)
+			reqBody.sort(
+				esb.sort(report._source.sorting[0][0], report._source.sorting[0][1])
+			);
+		else
+			reqBody.sort(
+				esb.sort(report._source.sorting[0], report._source.sorting[1])
+			);
+	}
+
+	//get the selected fields only
+	if (report._source.fields_exist) {
+		reqBody.source({ includes: report._source.selectedFields });
+	}
+	return reqBody;
+};
+
+// Fetch the data from ES
+export const getEsData = (arrayHits, report) => {
+	let hits: any = [];
+	for (let valueRes of arrayHits) {
+		for (let data of valueRes.hits) {
+			const fields = data.fields;
+			//get  all the fields of type date and fromat them to excel format
+			for (let dateType of report._source.dateFields) {
+				if (data._source[dateType]) {
```

Comment:
Is there a way to optimize this 3-layer for loop?

Reply:
I am using 3-layer for loop because of the the Elasticsearch data response structure.